File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:

    # 이익이 0보다 클 경우 최대값 - 최소값
    def maxProfit(self, prices: List[int]) -> int:
        non = 0
        take = prices[0]
        for i in range(1, len(prices)):
            # 점화식을 먼저 세우자
            price = prices[i]

            non = max(non, price - take)
            take = min(take, price)
        
        return non

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
/**

    현재 주식 가격에서 뒤에 있는 주식 가격의 최대값 만큼의 차액이 최대 이익이 된다.

    [1,2,3,4] i=0 -> arr[3] - arr[i] = 3

    [1,5,2,3] arr[1] - arr[0] = 4

 */

class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int maxProfit = 0;
        int maxPrice = 0;

        for (int i = n-1; i > -1; i--) {
            int price = prices[i];

            maxProfit = Math.max(maxProfit, maxPrice - price);
            maxPrice = Math.max(maxPrice, price);    
        }

        return maxProfit;
    }
}
